rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,697 @@
+
+        fixture.debug_print_files().await.unwrap();
+
+        let after_count = fixture.count_files().await.unwrap();
+        assert_eq!(after_count.num_index_files, 0);
+        assert_eq!(after_count.num_data_files, 1);
+        assert_eq!(after_count.num_manifest_files, 2);
+    }
+
+    #[tokio::test]
+    async fn clean_old_delete_files() {
+        let fixture = MockDatasetFixture::try_new().unwrap();
+        let mut data_gen = BatchGenerator::new().col(Box::new(
+            IncrementingInt32::new().named("filter_me".to_owned()),
+        ));
+
+        fixture
+            .create_with_data(data_gen.batch(16).unwrap())
+            .await
+            .unwrap();
+        fixture
+            .append_data(data_gen.batch(16).unwrap())
+            .await
+            .unwrap();
+        // This will keep some data from the appended file and should
+        // completely remove the first file
+        fixture.delete_data("filter_me < 20").await.unwrap();
+        fixture.clock.set_system_time(Duration::days(10));
+        fixture
+            .overwrite_data(data_gen.batch(16).unwrap())
+            .await
+            .unwrap();
+        // This will delete half of the last fragment
+        fixture.delete_data("filter_me >= 40").await.unwrap();
+
+        let before_count = fixture.count_files().await.unwrap();
+        assert_eq!(before_count.num_data_files, 3);
+        assert_eq!(before_count.num_delete_files, 2);
+        assert_eq!(before_count.num_manifest_files, 6);
+
+        let before = utc_now() - Duration::days(8);
+        fixture.run_cleanup(before).await.unwrap();
+
+        let after_count = fixture.count_files().await.unwrap();
+        assert_eq!(after_count.num_data_files, 1);
+        assert_eq!(after_count.num_delete_files, 1);
+        assert_eq!(after_count.num_manifest_files, 3);
+
+        // Ensure we can still read the dataset
+        let row_count_after = fixture.count_rows().await.unwrap();
+        assert_eq!(row_count_after, 8);
+    }
+
+    #[tokio::test]
+    async fn dont_clean_index_data_files() {
+        // Indexes have .lance files in them that are not referenced
+        // by any fragment.  We need to make sure the cleanup routine
+        // doesn't over-zealously delete these
+        let fixture = MockDatasetFixture::try_new().unwrap();
+        fixture.clock.set_system_time(Duration::days(10));
+        fixture.create_some_data().await.unwrap();
+        fixture.create_some_index().await.unwrap();
+
+        fixture.debug_print_files().await.unwrap();
+
+        let before_count = fixture.count_files().await.unwrap();
+        let before = utc_now() - Duration::days(8);
+        fixture.run_cleanup(before).await.unwrap();
+        let after_count = fixture.count_files().await.unwrap();
+
+        fixture.debug_print_files().await.unwrap();
+
+        assert_eq!(before_count, after_count);
+    }
+
+    #[tokio::test]
+    async fn cleanup_failed_commit_data_file() {
```

Comment:
Do we test that we don't cleanup files from in-progress writes? And where do we handle that in the implementation?

Reply:
Ah, good quesiton.  We mandated the `before`parameter had to be at least 7 days old but I still need to check the timestamp of the data files.  I'll add a test for this.

Reply:
I've added a new test for this and this gets handled by never deleting files that have been modified after `before`.